// openclaw/dist/plugin-sdk/synology-chat.js
import { h as DEFAULT_ACCOUNT_ID } from "./session-key-CbP51u9x.js";
import "./paths-WR8OhEmw.js";
import "./logger-s5D1BfzX.js";
import "./tmp-openclaw-dir-DEAexD45.js";
import "./subsystem-7GlwMGJM.js";
import "./utils-B2utBG_m.js";
import { H as setAccountEnabledInConfigSection } from "./zod-schema.core-CtLVNGPW.js";
import { t as registerPluginHttpRoute } from "./http-registry-7Lsnrxx9.js";
import { a as isRequestBodyLimitError, c as requestBodyErrorToText, s as readRequestBodyWithLimit } from "./http-body-CdacFz6y.js";
import { t as emptyPluginConfigSchema } from "./config-schema-D5FeRBQU.js";
import { i as buildChannelConfigSchema } from "./config-schema-HN6oil5E.js";
import { a as createFixedWindowRateLimiter } from "./webhook-memory-guards-CE-6mYdx.js";
export { DEFAULT_ACCOUNT_ID, buildChannelConfigSchema, createFixedWindowRateLimiter, emptyPluginConfigSchema, isRequestBodyLimitError, readRequestBodyWithLimit, registerPluginHttpRoute, requestBodyErrorToText, setAccountEnabledInConfigSection };
