// openclaw/dist/plugin-sdk/voice-call.js
import "./github-copilot-token-C4Aasre6.js";
import "./channel-config-helpers-BAi2qlAE.js";
import { _i as resolveOpenAITtsInstructions } from "./thread-bindings-SYAnWHuW.js";
import "./paths-WR8OhEmw.js";
import "./logger-s5D1BfzX.js";
import "./tmp-openclaw-dir-DEAexD45.js";
import "./subsystem-7GlwMGJM.js";
import { y as sleep } from "./utils-B2utBG_m.js";
import "./fetch-CtE9nADJ.js";
import "./exec-CbLkwGIT.js";
import "./thinking-D7ZEZ42s.js";
import "./query-expansion-DrxgDYz1.js";
import "./logger-DD_BJ94Q.js";
import { A as TtsConfigSchema, M as TtsProviderSchema, j as TtsModeSchema, k as TtsAutoSchema } from "./zod-schema.core-CtLVNGPW.js";
import "./redact-HAC0uPMW.js";
import "./http-registry-7Lsnrxx9.js";
import "./pairing-token-DNSESXf3.js";
import "./ssrf-BC5-OCfy.js";
import { t as fetchWithSsrFGuard } from "./fetch-guard-RV5sCukz.js";
import "./registry-B_tf98QY.js";
import { a as isRequestBodyLimitError, c as requestBodyErrorToText, s as readRequestBodyWithLimit } from "./http-body-CdacFz6y.js";
export { TtsAutoSchema, TtsConfigSchema, TtsModeSchema, TtsProviderSchema, fetchWithSsrFGuard, isRequestBodyLimitError, readRequestBodyWithLimit, requestBodyErrorToText, resolveOpenAITtsInstructions, sleep };
