// openclaw/dist/plugin-sdk/web-0jSlf-ol.js
import "./github-copilot-token-C4Aasre6.js";
import { C as pickWebChannel, b as logWebSelfId, k as webAuthExists, v as WA_WEB_AUTH_DIR } from "./channel-config-helpers-BAi2qlAE.js";
import "./thread-bindings-SYAnWHuW.js";
import "./paths-WR8OhEmw.js";
import "./logger-s5D1BfzX.js";
import "./tmp-openclaw-dir-DEAexD45.js";
import "./subsystem-7GlwMGJM.js";
import "./utils-B2utBG_m.js";
import "./fetch-CtE9nADJ.js";
import "./exec-CbLkwGIT.js";
import "./thinking-D7ZEZ42s.js";
import "./query-expansion-DrxgDYz1.js";
import "./logger-DD_BJ94Q.js";
import "./zod-schema.core-CtLVNGPW.js";
import "./redact-HAC0uPMW.js";
import "./http-registry-7Lsnrxx9.js";
import "./pairing-token-DNSESXf3.js";
import "./ssrf-BC5-OCfy.js";
import "./fetch-guard-RV5sCukz.js";
import "./registry-B_tf98QY.js";
import "./http-body-CdacFz6y.js";
import { n as monitorWebInbox, t as monitorWebChannel } from "./channel-web-DXX6S0Of.js";
import { t as sendMessageWhatsApp } from "./outbound-h4APtBc_.js";
import { i as waitForWaConnection, t as createWaSocket } from "./session-CAP3QgN3.js";
import { t as loginWeb } from "./login-uOFLI821.js";
export { WA_WEB_AUTH_DIR, createWaSocket, logWebSelfId, loginWeb, monitorWebChannel, monitorWebInbox, pickWebChannel, sendMessageWhatsApp, waitForWaConnection, webAuthExists };
