// openclaw/dist/plugin-sdk/whatsapp.js
import "./github-copilot-token-C4Aasre6.js";
import { g as normalizeAccountId, h as DEFAULT_ACCOUNT_ID } from "./session-key-CbP51u9x.js";
import { F as normalizeWhatsAppMessagingTarget, N as looksLikeWhatsAppTargetId, V as formatPairingApproveHint, a as formatWhatsAppConfigAllowFromEntries, d as resolveWhatsAppConfigDefaultTo, h as resolveWhatsAppAccount, m as resolveDefaultWhatsAppAccountId, p as listWhatsAppAccountIds, u as resolveWhatsAppConfigAllowFrom } from "./channel-config-helpers-BAi2qlAE.js";
import { Da as listWhatsAppDirectoryGroupsFromConfig, Fa as createWhatsAppOutboundBase, Ia as resolveWhatsAppGroupIntroHint, Jr as createActionGate, La as resolveWhatsAppMentionStripPatterns, Oa as listWhatsAppDirectoryPeersFromConfig, Qr as readStringParam, Ra as resolveWhatsAppOutboundTarget, bs as getChatChannelMeta, eo as resolveWhatsAppGroupRequireMention, s as WhatsAppConfigSchema, to as resolveWhatsAppGroupToolPolicy } from "./thread-bindings-SYAnWHuW.js";
import "./paths-WR8OhEmw.js";
import "./logger-s5D1BfzX.js";
import "./tmp-openclaw-dir-DEAexD45.js";
import "./subsystem-7GlwMGJM.js";
import { d as normalizeE164 } from "./utils-B2utBG_m.js";
import "./fetch-CtE9nADJ.js";
import "./exec-CbLkwGIT.js";
import "./thinking-D7ZEZ42s.js";
import "./query-expansion-DrxgDYz1.js";
import "./logger-DD_BJ94Q.js";
import "./zod-schema.core-CtLVNGPW.js";
import "./redact-HAC0uPMW.js";
import "./http-registry-7Lsnrxx9.js";
import "./pairing-token-DNSESXf3.js";
import "./ssrf-BC5-OCfy.js";
import "./fetch-guard-RV5sCukz.js";
import { i as resolveDefaultGroupPolicy, r as resolveAllowlistProviderRuntimeGroupPolicy } from "./runtime-group-policy-KevvtwrL.js";
import "./registry-B_tf98QY.js";
import "./http-body-CdacFz6y.js";
import { t as emptyPluginConfigSchema } from "./config-schema-D5FeRBQU.js";
import "./provider-env-vars-DNcS2Y9_.js";
import { D as applyAccountNameToChannelSection, k as migrateBaseNameToDefaultAccount } from "./helpers-uRGq4pbi.js";
import { i as buildChannelConfigSchema } from "./config-schema-HN6oil5E.js";
import "./shared-Br_iEHnn.js";
import { r as resolveWhatsAppHeartbeatRecipients } from "./channel-web-DXX6S0Of.js";
import "./outbound-h4APtBc_.js";
import "./session-CAP3QgN3.js";
import "./login-uOFLI821.js";
import { n as whatsappOnboardingAdapter, t as collectWhatsAppStatusIssues } from "./whatsapp-BUh31yu3.js";
export { DEFAULT_ACCOUNT_ID, WhatsAppConfigSchema, applyAccountNameToChannelSection, buildChannelConfigSchema, collectWhatsAppStatusIssues, createActionGate, createWhatsAppOutboundBase, emptyPluginConfigSchema, formatPairingApproveHint, formatWhatsAppConfigAllowFromEntries, getChatChannelMeta, listWhatsAppAccountIds, listWhatsAppDirectoryGroupsFromConfig, listWhatsAppDirectoryPeersFromConfig, looksLikeWhatsAppTargetId, migrateBaseNameToDefaultAccount, normalizeAccountId, normalizeE164, normalizeWhatsAppMessagingTarget, readStringParam, resolveAllowlistProviderRuntimeGroupPolicy, resolveDefaultGroupPolicy, resolveDefaultWhatsAppAccountId, resolveWhatsAppAccount, resolveWhatsAppConfigAllowFrom, resolveWhatsAppConfigDefaultTo, resolveWhatsAppGroupIntroHint, resolveWhatsAppGroupRequireMention, resolveWhatsAppGroupToolPolicy, resolveWhatsAppHeartbeatRecipients, resolveWhatsAppMentionStripPatterns, resolveWhatsAppOutboundTarget, whatsappOnboardingAdapter };
